File: cpp/questions/spiralOrder.cpp
#include <vector>

using namespace std;

class Solution {
private:
    static constexpr int DIR[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int rs = matrix.size();
        int cs = matrix[0].size();
        std::vector<int> ans(rs*cs);
        int l = 0;
        int ac = 0;
        int d = 0;
        int r = 0;
        int c = 0;
        while (ac < rs*cs) {
            ans[ac++] = matrix[r][c];
            switch (d) {
                case 0:
                    if (c == cs-l-1) {
                        d = 1;
                    }
                    break;
                case 1:
                    if (r == rs-l-1) {
                        d = 2;
                    }
                    break;
                case 2:
                    if (c == l) {
                        d = 3;
                    }
                    break;
                case 3:
                    if (r == l+1) {
                        d = 0;
                        l += 1;
                    }
                    break;
            }
            r += DIR[d][0];
            c += DIR[d][1];
        }
        return ans;
    }
};

File: python/questions/spiralOrder.py
from typing import List

class Solution:
    DIR = [[0, 1], [1, 0], [0, -1], [-1, 0]]

    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rs = len(matrix)
        cs = len(matrix[0])
        ans = [0] * rs * cs
        d = 0
        r = 0
        c = 0
        oc = 0
        l = 0
        while oc < rs*cs:
            ans[oc] = matrix[r][c]
            oc += 1
            match d:
                case 0:
                    if c == cs-1-l:
                        d = 1
                case 1:
                    if r == rs-1-l:
                        d = 2
                case 2:
                    if c == l:
                        d = 3
                case 3:
                    if r == l+1:
                        l += 1
                        d = 0
            r += self.DIR[d][0]
            c += self.DIR[d][1]
        return ans
